### PR TITLE
Implementation Series.cov in new style

### PR DESCRIPTION
**Pull request by Rubtsowa**

### CODE REVIEW

File: hpat/hiframes/hiframes_typed.py [hunk 1]
```diff
@@ -972,7 +972,7 @@ def run_call_series_quantile_default(A):
             return self._replace_func(
                 func, (data, index, n_arg, name), pre_nodes=nodes)
 
-        if func_name in ('cov', 'corr'):
+        if func_name in ('corr'):
```
Comment (densmirn) [inline hpat/hiframes/hiframes_typed.py]: Actually `('corr')` is not tuple, it's string. So the condition is not so correct.

File: hpat/hiframes/hiframes_typed.py [hunk 2]
```diff
@@ -1507,15 +1507,15 @@ def _run_call_series_rolling(self, assign, lhs, rhs, rolling_var, func_name):
             nodes.append(ir.Assign(ir.Const(center, lhs.loc), center_var, lhs.loc))
             center = center_var
 
-        if func_name in ('cov', 'corr'):
+        if func_name in ('corr'):
```
Comment (densmirn) [inline hpat/hiframes/hiframes_typed.py]: I'm not sure the change is required, because it related to some rolling.

File: hpat/hiframes/hiframes_typed.py [hunk 3]
```diff
@@ -1507,15 +1507,15 @@ def _run_call_series_rolling(self, assign, lhs, rhs, rolling_var, func_name):
...
             if len(rhs.args) == 1:
                 other = self._get_series_data(rhs.args[0], nodes)
             else:
                 other = data
-            if func_name == 'cov':
-                def f(a, b, w, c):
-                    return hpat.hiframes.api.init_series(hpat.hiframes.rolling.rolling_cov(a, b, w, c))
+            # if func_name == 'cov':
+            #     def f(a, b, w, c):
+            #         return hpat.hiframes.api.init_series(hpat.hiframes.rolling.rolling_cov(a, b, w, c))
```
Comment (densmirn) [inline hpat/hiframes/hiframes_typed.py]: The same, perhaps not required.

File: hpat/hiframes/pd_series_ext.py [hunk 4]
```diff
@@ -847,9 +847,9 @@ def resolve_apply(self, ary, args, kws):
         # result is always float64 (see Pandas window.pyx:roll_generic)
         return signature(SeriesType(types.float64), *args)
 
-    @bound_function("rolling.cov")
-    def resolve_cov(self, ary, args, kws):
-        return signature(SeriesType(types.float64), *args)
+    # @bound_function("rolling.cov")
+    # def resolve_cov(self, ary, args, kws):
+    #     return signature(SeriesType(types.float64), *args)
```
Comment (densmirn) [inline hpat/hiframes/pd_series_ext.py]: The same, the change is not required. It's not related to pd.Series.

File: hpat/hiframes/pd_series_ext.py [hunk 5]
```diff
@@ -988,7 +988,7 @@ def generic_expand_cumulative_series(self, args, kws):
 
 # TODO: add itemsize, strides, etc. when removed from Pandas
 _not_series_array_attrs = ['flat', 'ctypes', 'itemset', 'reshape', 'sort', 'flatten',
-                           'resolve_take', 'resolve_max', 'resolve_min', 'resolve_nunique']
+                           'resolve_take', 'resolve_max', 'resolve_min', 'resolve_nunique', 'resolve_cov']
```
Comment (densmirn) [inline hpat/hiframes/pd_series_ext.py]: `'resolve_cov'` could be inserted in alphabetical order.

File: hpat/hiframes/series_kernels.py [hunk 6]
```diff
@@ -514,7 +514,7 @@ def gt_f(a, b):
     'str_contains_regex': _str_contains_regex_impl,
     'str_contains_noregex': _str_contains_noregex_impl,
     # 'abs': lambda A: hpat.hiframes.api.init_series(np.abs(A)),  # TODO: timedelta
-    'cov': _column_cov_impl,
+    #'cov': _column_cov_impl,
```
Comment (densmirn) [inline hpat/hiframes/series_kernels.py]: The change is not required.

File: hpat/tests/test_series.py [hunk 7]
```diff
@@ -186,14 +186,22 @@ def test_impl(A):
         S = pd.Series([3, 5, 6], ['a', 'b', 'c'], name='A')
         pd.testing.assert_series_equal(hpat_func(S), test_impl(S))
 
-    def test_series_attr1(self):
```
Comment (densmirn) [inline hpat/tests/test_series.py]: The changes shouldn't be in this PR.

File: hpat/tests/test_series.py [hunk 8]
```diff
@@ -249,6 +257,26 @@ def test_impl(A):
         df = pd.DataFrame({'A': np.arange(n)})
         np.testing.assert_array_equal(hpat_func(df.A), test_impl(df.A))
 
+    def test_series_getattr_ndim(self):
```
Comment (densmirn) [inline hpat/tests/test_series.py]: The same, the changes shouldn't be in this PR.

File: hpat/tests/test_series.py [hunk 9]
```diff
@@ -303,6 +331,32 @@ def test_impl(S):
         S = pd.Series(['aa', 'bb', 'cc'])
         pd.testing.assert_series_equal(hpat_func(S), test_impl(S))
 
+    def test_series_astype_str_to_str_index_str(self):
```
Comment (densmirn) [inline hpat/tests/test_series.py]: The same, the changes shouldn't be in this PR.

File: hpat/tests/test_series.py [hunk 10]
```diff
@@ -2099,6 +2150,62 @@ def test_series_nunique_param1_impl(S, dropna):
                     result_param1 = hpat_func_param1(S, param1)
                     self.assertEqual(result_param1, result_param1_ref)
 
+    def test_series_cov1(self):
```
Comment (densmirn) [inline hpat/tests/test_series.py]: Let's think up better name for the tests related to what happens in the test.

File: hpat/tests/test_series.py [hunk 11]
```diff
@@ -2099,6 +2150,62 @@ def test_series_nunique_param1_impl(S, dropna):
...
+                    np.testing.assert_allclose(result, result_ref)
+
+    def test_series_cov2(self):
+        def test_series_cov_impl(S1, S2, min_periods=None):
+            return S1.cov(S2)
+
+        hpat_func = hpat.jit(test_series_cov_impl)
+        S1 = pd.Series([.2, .0, .6, .2])
+        S2 = pd.Series(['abcdefgh', 'a','abcdefg', 'ab', 'abcdef', 'abc'])
+        with self.assertRaises(TypeError):
```
Comment (densmirn) [inline hpat/tests/test_series.py]: It would be better to check error message from exception via `self.assertIn`.

File: hpat/hiframes/hiframes_typed.py (hunk 1 above)
Comment (PokhodenkoSA) [inline hpat/hiframes/hiframes_typed.py]:
```suggestion
        if func_name in ('corr',):
```

File: hpat/hiframes/pd_series_ext.py (hunk 5 above)
Comment (PokhodenkoSA) [inline hpat/hiframes/pd_series_ext.py]:
```suggestion
                           'resolve_cov',
                           'resolve_take', 'resolve_max', 'resolve_min', 'resolve_nunique']
```

File: hpat/tests/test_series.py [hunk 12]
```diff
@@ -186,14 +186,22 @@ def test_impl(A):
...
+        for S, expected in [
+            (pd.Series(), 0),
+            (pd.Series([]), 0),
+            (pd.Series(np.arange(n)), n),
+            (pd.Series([np.nan, 1, 2]), 3),
+            (pd.Series(['1', '2', '3']), 3),
+        ]:
+            with self.subTest(S=S, expected=expected):
+                self.assertEqual(hpat_func(S), expected)
+                self.assertEqual(hpat_func(S), test_impl(S))
```
Comment (PokhodenkoSA) [inline hpat/tests/test_series.py]: This cod is in master. I think your branch should be merged/rebased with master.

File: hpat/tests/test_series.py (hunk 8 above)
Comment (PokhodenkoSA) [reply hpat/tests/test_series.py]: I think it is master too.

File: hpat/datatypes/hpat_pandas_series_functions.py [hunk 13]
```diff
@@ -1953,3 +1953,63 @@ def hpat_pandas_series_nunique_impl(self, dropna=True):
...
+
+    Returns
+    -------
+    :obj:`float`
+         returns :obj:`float` object
+    """
+
+    _func_name = 'Method cov().'
+
+    if not (isinstance(self, SeriesType) and isinstance(other, SeriesType)):
```
Comment (densmirn) [inline hpat/datatypes/hpat_pandas_series_functions.py]: I would split the condition on 2 ones.

File: hpat/datatypes/hpat_pandas_series_functions.py [hunk 14]
```diff
@@ -1953,3 +1953,63 @@ def hpat_pandas_series_nunique_impl(self, dropna=True):
...
+        raise TypingError('{} The object must be a pandas.series. Given: {} and {}'.format(_func_name, self, other))
+
+    if not (isinstance(self.data.dtype, types.Number) and isinstance(other.data.dtype, types.Number)):
```
Comment (densmirn) [inline hpat/datatypes/hpat_pandas_series_functions.py]: The same, I would split the condition on 2 ones.

File: hpat/datatypes/hpat_pandas_series_functions.py [hunk 15]
```diff
@@ -1953,3 +1953,63 @@ def hpat_pandas_series_nunique_impl(self, dropna=True):
...
+        raise TypingError('Currently function supports only numeric values. Given: {} and {}'.format(_func_name, self, other))
+
+    if not isinstance(min_periods, (types.Number, types.Omitted, types.NoneType)):
```
Comment (densmirn) [inline hpat/datatypes/hpat_pandas_series_functions.py]: `types.Number` -> `types.Integer`

File: hpat/hiframes/hiframes_typed.py [hunk 16]
```diff
@@ -1498,7 +1498,7 @@ def _run_call_series_rolling(self, assign, lhs, rhs, rolling_var, func_name):
             nodes.append(ir.Assign(ir.Const(center, lhs.loc), center_var, lhs.loc))
             center = center_var
 
-        if func_name in ('cov', 'corr'):
+        if func_name == 'corr':
```
Comment (densmirn) [inline hpat/hiframes/hiframes_typed.py]: It's related to rolling, please revert the change.

File: hpat/tests/test_series.py (hunk 11 above)
Comment (densmirn) [reply hpat/tests/test_series.py]: Please use `self.assertIn` where you catch exceptions.
